src/transformers/models/cohere/modeling_cohere.py

```diff
@@ -996,45 +1003,46 @@ def _update_causal_mask(
             if AttentionMaskConverter._ignore_causal_mask_sdpa(
-                attention_mask, inputs_embeds=input_tensor, past_key_values_length=past_seen_tokens
+                attention_mask,
+                inputs_embeds=input_tensor,
+                past_key_values_length=past_seen_tokens,
+                static_cache_max_length=static_cache_max_length,
             ):
                 return None
 
         dtype, device = input_tensor.dtype, input_tensor.device
         min_dtype = torch.finfo(dtype).min
         sequence_length = input_tensor.shape[1]
-        if hasattr(getattr(self.layers[0], "self_attn", {}), "past_key_value"):  # static cache
-            target_length = self.config.max_position_embeddings
-        else:  # dynamic cache
-            target_length = (
-                attention_mask.shape[-1]
-                if isinstance(attention_mask, torch.Tensor)
-                else past_seen_tokens + sequence_length + 1
-            )
 
-        causal_mask = torch.full((sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device)
-        if sequence_length != 1:
-            causal_mask = torch.triu(causal_mask, diagonal=1)
-        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
-        causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
-        if attention_mask is not None:
-            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
-            if attention_mask.dim() == 2:
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # we can pass both the full 4D mask (i.e. [..., full_len, full_len]) and a 4D mask with the same shape
+            # as the causal mask (i.e. [..., seq_len, full_len])
+            mask_slice = (attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype
+            offset = cache_position[0]
+            if attention_mask.shape[-2] == offset + sequence_length:
+                mask_slice = mask_slice[..., offset:, :]
+            causal_mask = mask_slice
+        else:
+            if hasattr(self.layers[0].self_attn, "past_key_value"):  # static cache
+                target_length = self.config.max_position_embeddings
+            else:  # dynamic cache
+                target_length = (
+                    attention_mask.shape[-1]
+                    if isinstance(attention_mask, torch.Tensor)
+                    else past_seen_tokens + sequence_length + 1
+                )
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
```

Comment:
this is very messy. 
IMO if a 4D mask is passed, then we should not even touch it. 
Previously we used to invert it, now that we don't do that in the process, we should not add extra logic and just return the 4D mask. Fine with me to add a check to make sure the min is -inf, max is 0 and shape is expected, but let's not increase the surface and complicate things.

Reply:
I simplified the code:
- removed inversion for custom 4d masks, 
- removed mask slicing for custom 4d masks
- removed test case where we pass a 4D attention mask with the full sequence length (i.e. [..., full_len, full_len]))
- updated relevant tests

I am concerned that 
- part of code related to checking masks shape sits in  `_ignore_causal_mask_sdpa()`
- the advanced 4d mask tests sit in `tests/models/llama/test_modeling_llama.py` - something may break them again outside of llama code, like in `src/transformers/modeling_attn_mask_utils.py` - could you please ensure that  `Mask4DTestHard` class is always tested if something changes in `modeling_attn_mask_utils.py` even if Llama code is not affected otherwise?